### PR TITLE
256 audit pdm 006 add weight charging to chain extension calls

### LINKED ISSUE #256: Add weight charging to chain extension calls
As pointed out by our auditors, with the current implementation of our chain extensions, smart contract calls can result in state changes without having to pay any weight fee. 

> By not charging weight for these operations, the system becomes vulnerable to potential security issues, particularly Denial-of-Service (DoS) attacks. Malicious contracts can exploit this lack of weight charging by flooding the system with a high volume of calls, overloading the network and disrupting its normal operation.

We can do that as follows, before accessing contract memory:
```
let charged_weight = env.charge_weight(sufficient_weight)?; 
trace!( 
    target: "runtime", 
    "[ChainExtension]|call|func_id / charge_weight:{:?}", 
    charged_weight 
)
```

We should add that logic to any chain extension invocation and not just the ones changing state though. Example implementations can be found in [this](https://github.com/AstarNetwork/Astar/blob/master/chain-extensions/dapps-staking/src/lib.rs) file and [these](https://github.com/AstarNetwork/Astar/blob/f105448a28b530dca6028c02e666b7999764959e/chain-extensions/dapps-staking/src/lib.rs#L191-L192) lines.

### PR DESCRIPTION
**Pull request by ashneverdawn**
Closes #256 

Chain extension `overhead` calculation is based on: https://github.com/paritytech/ink/blob/255fc20c69cb7c12ff70c4fb0aef190d948deba2/integration-tests/psp22-extension/runtime/psp22-extension-example.rs#L233-L246

It uses `from_ref_time` which is deprecated and issues a warning to use `from_parts` instead, to that's what I did.

I also had to expose `approve` and `transfer_from` as extrinsics to be able to create benchmarks for them and calculate their weights.

### CODE REVIEW

**Review by ebma — COMMENTED** [review]
It looks good to me overall. Maybe we should remove the `trace!()` statements as they clutter the code quite a lot. These statements were suggested in a snippet provided by the auditors but maybe it's not worth keeping them.

At the moment, the benchmarks assume the light-weight scenario, ie. a vector of length 1. As pointed out [here](https://github.com/pendulum-chain/pendulum/issues/257#issuecomment-1624703914), we could and should instead write benchmarks that create weights that take the length of that vector into account. We can do that in another PR targeting #257 explicitly though.

File: runtime/foucoco/src/lib.rs [hunk 1]
```diff
@@ -1085,7 +1130,16 @@ where
 				let ext = env.ext();
 				let caller = ext.caller().clone();
 
-				let env = env.buf_in_buf_out();
+				let mut env = env.buf_in_buf_out();
+				let base_weight =
+					<<T as orml_currencies_allowance_extension::Config>::WeightInfo as orml_currencies_allowance_extension::WeightInfo>::approve();
```
[the remark this answers is not in the document]
Comment (ashneverdawn) [reply runtime/foucoco/src/lib.rs]: Unfortunately I can't assign it to a variable without getting this compile error:
```
Syntax Error: expected `::`
```
Maybe it's possible with a trait alias, but couldn't figure that one out either. I ended up just adding a `use` for the traits, which made it considerably shorter.

[the remark this answers is not in the document]
Comment (ashneverdawn) [reply runtime/foucoco/src/lib.rs]: I agree about another PR targeting https://github.com/pendulum-chain/pendulum/issues/257 explicitly 👍

[the remark this answers is not in the document]
Comment (ashneverdawn) [reply runtime/foucoco/src/lib.rs]: I removed the `trace!()` statements 👍

[the remark this answers is not in the document]
Comment (ebma) [reply runtime/foucoco/src/lib.rs]: Nice, the code looks much cleaner now  👍

**Review by ebma — APPROVED** [review]
Thanks for adding my requested changes @ashneverdawn 👍 
LGTM now and ready for merge.

**Pull request merged**